### PR TITLE
feat: add EIP-1898 blockHash option to JSON-RPC methods

### PR DESCRIPTION
**Pull request by ahonn**
**What this PR does / why we need it**:

https://eips.ethereum.org/EIPS/eip-1898

**Which issue(s) this PR fixes**:

Fixes #

**Which docs this PR relation**:

Ref #

**Which toolchain this PR adaption**:

No Breaking Change

**Special notes for your reviewer**:

NIL

**CI Description**

| CI Name                                   | Description                                                     |
| ----------------------------------------- | --------------------------------------------------------------- |
| *Chaos CI*                                | Test the liveness and robustness of Axon under terrible network condition    |
| *Cargo Clippy*                            | Run `cargo clippy --all --all-targets --all-features`      |
| *Coverage Test*                           | Get the unit test coverage report                             |
| *E2E Test*                                | Run end-to-end test to check interfaces                         |
| *Code Format*                             | Run `cargo +nightly fmt --all -- --check` and `cargo sort -gwc`     |
| *Web3 Compatible Test*                    | Test the Web3 compatibility of Axon                               |
| *v3 Core Test*                            | Run the compatibility tests provided by Uniswap V3             |
| *OCT 1-5 \| 6-10 \| 11 \| 12-15 \| 16-19* | Run the compatibility tests provided by OpenZeppelin           |

**CI Usage**

> Check the CI you want to run below, and then comment `/run-ci`.

**CI Switch**

- [ ] Chaos CI
- [x] Cargo Clippy
- [ ] Coverage Test
- [x] E2E Tests
- [x] Code Format
- [x] Unit Tests
- [x] Web3 Compatible Tests
- [x] OCT 1-5 And 12-15
- [x] OCT 6-10
- [x] OCT 11
- [x] OCT 16-19
- [x] v3 Core Tests

### CODE REVIEW

File: core/api/src/jsonrpc/web3_types.rs [hunk 1]
```diff
@@ -336,6 +336,51 @@ pub struct Web3CallRequest {
...
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
+#[serde(deny_unknown_fields, rename_all = "camelCase")]
+pub struct BlockSpecifier {
+  #[serde(skip_serializing_if = "Option::is_none")]
+  pub block_number: Option<u64>,
+  #[serde(skip_serializing_if = "Option::is_none")]
+  pub block_hash: Option<H256>,
+  #[serde(skip_serializing_if = "Option::is_none")]
+  pub require_canonical: Option<bool>,
```
Comment (KaoImin) [inline core/api/src/jsonrpc/web3_types.rs]: Axon is based BFT consensus which never fork.Whether the `require_canonical` is true or not leads to the same process. So do not need to parse this field.

**Review by KaoImin — CHANGES_REQUESTED** [review]
Only need to extend the `BlockId` enumeration like this:
```rust
#[derive(Default, Clone, Debug, PartialEq, Eq, Hash)]
pub enum BlockId {
    Num(u64),
    Hash(H256),
    #[default]
    Latest,
    Earliest,
    Pending,
}
```

File: core/api/src/jsonrpc/web3_types.rs [hunk 2]
```diff
@@ -339,6 +339,7 @@ pub struct Web3CallRequest {
 #[derive(Default, Clone, Debug, PartialEq, Eq, Hash)]
 pub enum BlockId {
     Num(u64),
```
Comment (KaoImin) [inline core/api/src/jsonrpc/web3_types.rs]:
```suggestion
    Num(U64),
```

File: core/api/src/jsonrpc/web3_types.rs [hunk 3]
```diff
@@ -371,6 +372,7 @@ impl Serialize for BlockId {
     {
         match *self {
             BlockId::Num(ref x) => serializer.serialize_str(&format!("0x{:x}", x)),
```
Comment (KaoImin) [inline core/api/src/jsonrpc/web3_types.rs]:
```suggestion
            BlockId::Num(ref x) => serializer.serialize_str(&format!("{:x}", x)),
```

**Review by KaoImin — COMMENTED** [review]
The type of `Num` in `BlockID` should be `U64`, the current code style is inelegant. It should be changed.

**Review by KaoImin — DISMISSED** [review]

File: core/api/src/adapter.rs [hunk 4]
```diff
@@ -100,6 +100,22 @@ where
         }
     }
 
+    async fn get_block_header_by_hash(
```
Comment (KaoImin) [inline core/api/src/adapter.rs]: It seems a useless interface.

Comment (ahonn) [reply core/api/src/adapter.rs]: Yes. This was an legacy implementation that is no longer in use and has been deleted.

**Review by KaoImin — APPROVED** [review]

**Pull request merged**